src/librustc_mir/transform/move_up_propagation.rs

```diff
@@ -0,0 +1,596 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::ty::TyCtxt;
+use rustc::mir::repr::*;
+use rustc::mir::transform::{MirPass, MirSource, Pass};
+use rustc::mir::visit::{Visitor, LvalueContext, MutVisitor};
+use std::collections::HashSet;
+use rustc_data_structures::tuple_slice::TupleSlice;
+use rustc_data_structures::bitvec::BitVector;
+use rustc_data_structures::indexed_vec::{IndexVec, Idx};
+
+pub struct MoveUpPropagation;
+
+impl Pass for MoveUpPropagation {}
+
+impl<'tcx> MirPass<'tcx> for MoveUpPropagation {
+    fn run_pass<'a>(&mut self,
+                    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    src: MirSource,
+                    mir: &mut Mir<'tcx>) {
+        // we only return when mir_opt_level > 1
+        match tcx.sess.opts.debugging_opts.mir_opt_level {
+            Some(0) |
+            Some(1) |
+            None => { return; },
+            _ => {}
+        };
+        let node_id = src.item_id();
+        let node_path = tcx.item_path_str(tcx.map.local_def_id(node_id));
+        debug!("move-up-propagation on {:?}", node_path);
+
+        let mut opt_counter = 0;
+        let mut dead_temps = BitVector::new(mir.temp_decls.len());
+        let mut dead_vars = BitVector::new(mir.var_decls.len());
+        while let Some((tmp, use_bb, use_idx, def_bb, def_idx)) = get_one_optimization(mir) {
```

Comment:
I would like to see more comments here. Ideally, we would explain what the optimization will do, like showing one example of the transformation, and labeling the various parts of it. Then we can have comments that link the code to that transformation.